lib/spack/spack/config.py

```diff
@@ -772,13 +817,40 @@ def _mark_internal(data, name):
     return d
 
 
-def _merge_yaml(dest, source):
+def type_of(path):
+    """Returns an instance of a type that will pass validation for path.
+
+    The instance is created by calling the constructor with no arguments.
+    If multiple types will satisfy validation for data at the configuration
+    path given, the priority order is ``list``, ``dict``, ``str``, ``bool``,
+    ``int``, ``float``.
+    """
```

Comment:
Is it not possible to extract this using `jsonschema`?  Can't we just look this up in the schema itself?  `jsonschema` has some query capabilities, at least for errors:

* https://python-jsonschema.readthedocs.io/en/latest/errors/

Can you use that on the first error?  Or is there a way to ask the validator about a path without having to try to validate everything?

Reply:
I don't believe so. It appears that the only things you can get out of the interface are information about the instance being validated or the path into the schema at which it failed.

We could potentially get the path and look at the raw schema at that path, but I don't think that helps us. We would then have to write a parser that could parse any arbitrary combination of `'OneOf': {'AnyOf': {'type': ...` and I think the possible complexity there is not worth trying to model.